openshift_tools/cloud/aws/ebs_snapshotter.py

```diff
@@ -324,42 +330,55 @@ def get_expired_snapshots(snapshots, hourly_backups, daily_backups, weekly_backu
         # We want to make sure we're only sending back 1 copy of each snapshot to trim.
         return list(set(snaps_to_trim))
 
-    @staticmethod
-    def get_volume_snapshots(volume, all_snapshots):
-        """ Filters out the snapshots for a given volume from the entire list of snapshots. """
-        # Get this volume's specific snapshots
-        vol_snaps = [s for s in all_snapshots if s.volume_id == volume.id]
-        EbsSnapshotter.sort_snapshots(vol_snaps)
-        return vol_snaps
 
     # Reason: disable pylint too-many-arguments because this is the API I want to expose.
     # Status: permanently disabled
     # pylint: disable=too-many-arguments
-    def trim_snapshots(self, hourly_backups, daily_backups, weekly_backups, monthly_backups, dry_run=False):
+    # pylint: disable=too-many-statements
+    def trim_snapshots(self, hourly_backups,
+                       daily_backups,
+                       weekly_backups,
+                       monthly_backups,
+                       delete_orphans_older_than,
+                       dry_run=False):
         """ Finds expired snapshots given the number of hourly, daily, weekly and monthly backups
             to keep, then removes them.
         """
 
         all_expired_snapshots = []
         deleted_snapshots = []
+        volume_snapshots = defaultdict(list)
+        orphan_snapshots = []
         errors = []
 
         volumes = self.get_volumes_with_schedule(ALL_SCHEDULES)
 
-        if len(volumes) == 0:
-            # Early exit for any region that doesn't have snapshotable volumes
-            return (all_expired_snapshots, deleted_snapshots, errors)
+        volume_ids = []
+        for volume in volumes:
+            volume_ids.append(volume.id)
+
+        # get a list of snapshots that we own from AWS
+        all_snapshots = self.ec2.get_all_snapshots(owner='self')
 
-        all_snapshots = self.ec2.get_all_snapshots()
+        # sort the snapshots into the following categories:
+        # - snapshots with a volume that currently exists
+        # - snapshots that are orphaned and have no volume anymore
+        # - ignore orphaned snapshots that we did not create, that have no 'snapshot' tag
+        for snapshot in all_snapshots:
+            if snapshot.volume_id in volume_ids:
+                volume_snapshots[snapshot.volume_id].append(snapshot)
+            else:
+                if 'snapshot' in snapshot.tags and snapshot.tags['snapshot'] in SUPPORTED_SCHEDULES:
```

Comment:
Add a comment to not remove ```snapshot.tags['snapshot'] in SUPPORTED_SCHEDULES``` as it's crucial for this to only remove snapshots that are part of this snapshotting / trimming system.